adding design by contract lesson

markcmiller86 opened the pull request:
I was on the fence about where best to add this. It most closely resembles the Assertions content. So, I added it as a 2a lesson closely following Assertions.

As an aside, I wasn't sure as to the rationale for numbering entries either in index.md or as the filenames. Is this some pattern enforced by Software Caprentry? As you can imagine, it makes for added headaches moving things around later due to need to either renumber (undesireable) or leading to confusing/out-of-order numbering. So, I opted for a '2a' numbering.

markcmiller86 commented:
After having a couple of days away from it, I updated the content a bit, making it a bit clearer, I hope.

katyhuff commented on index.md:
```diff
@@ -20,6 +20,7 @@ title: Testing
 
 1.  [Basics of Testing](01-basics.html)
 2.  [Assertions](02-assertions.html)
+2a. [Design By Contract](02a-dbc.html)
```
This does not render properly in markdown.

![image](https://cloud.githubusercontent.com/assets/393899/25724050/04485fa0-30e0-11e7-9f52-98630a12423d.png)

Maybe this section would be appropriate after exceptions or after unit tests anyway?

markcmiller86 replied:
FWIW, I felt the rendering of 2a on the same line as 2 _was_ correct.

And, part of my issue in placing it at same levels as others was that it forces a re-numbering of everything there which I felt was intrusive (for you) not to mention tedious (for me).

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+> ## Learning Objectives {.objectives}
+>
+> *   Design by Contract is a way of using Assertions for interface specification.
+> *   Design by Contract defines pre-, post- and invariant conditions a function and a caller agree to obey.
```
This objective might be more parsable to a beginner if it's more granular. For example, understanding the meaning of "pre-, post- and invariant conditions" in this context could be its own objective.

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+* Post-conditions: Things that are guaranteed to be true after the function returns.
+* Invariant-conditions: Things that are guaranteed not to change during the function's execution
+
+**Note**: In Python's contracts, only pre- and post-conditions are handled.
```
perhaps just "In Python contracts" ?

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+
+**Note**: In Python's contracts, only pre- and post-conditions are handled.
+
+In the examples here, we use [PyContracts](https://andreacensi.github.io/contracts/index.html#) which use
```
Since "PyContracts" is just one library, I think this should be "which uses". The fact that PyContracts ends in an s doesn't make it a plural noun, really, in this case.

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+~~~
+
+To demonstrate the use of contracts, in the example here, we implement our own version of an integer square root
+function for perfect squares, called psqrt. We define a contract that indicates the caller is required to pass an
```
The name of this function (psqrt) perhaps could be formatted like code, so that it stands out (like ``` `psqrt` ```)

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+> *   Design by Contract is a way of using Assertions for interface specification.
+> *   Design by Contract defines pre-, post- and invariant conditions a function and a caller agree to obey.
+
+In [Design by Contract](https://en.wikipedia.org/wiki/Design_by_contract), the interaction between a caller
```
should define "caller"

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+
+Values of 4 and 81 are both integers. So, in these cases the caller has obeyed the pre-conditions of the contract.
+In addition, because both 4 and 81 are perfect squares, the function correctly returns their integer square root.
+So, the funtion has obeyd the post-conditions of the contract.
```
obeyd -> obeyed

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+In addition, because both 4 and 81 are perfect squares, the function correctly returns their integer square root.
+So, the funtion has obeyd the post-conditions of the contract.
+
+Now, lets see what happens when the caller failes to obey the pre-conditions of the contract by passing a negative
```
failes -> fails

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+Variables bound in inner context:
+~~~
+
+An exception is raised indicating a failure to obey the pre-condition for passing a value greather than or equal to zero.
```
at this point, the students haven't seen exceptions. That is in the next part of the lesson (3).

markcmiller86 replied:
Well, I guess that is more motivation to move to section 3a (or 4 except for the ensuing re-numbering).

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+
+### Extending Contracts
+
+Sometimes, the simple builtin syntax for defining contracts is not sufficient. In this case, contracts can be extended by
```
perhaps "built-in" ?

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+
+~~~ {.python}
+@new_contract
+def endsOk(x):
```
this function name is a bit odd. can I recommend a more pep8 friendly ends_okay() ?

markcmiller86 replied:
how about ends_ok...ends_okay is too many chars longer, requires reformatting of error text too.

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+
+~~~ {.python}
+@contract(x='int,endsOk,>=0',returns='int,>=0')
+def psqrt2(x):
```
maybe a more meaningful name could be put here. It probably won't be obvious to a beginner why this extra function is being written.

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+checking: int,endsOk,>=0   for value: Instance of <type 'int'>: 1000   
+Variables bound in inner context:
+~~~
+
```
This is a good spot to explain the code above. Did we see what we expect (an exception). Why did we expect that?

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+
+### Performance Considerations
+
+Depending on the situation, checking validity of a contract can be expensive. For example, suppose a function
```
the validity

markcmiller86 replied:
Hmm. The expense is associated with _checking_ a contract's validity, not with _the validity_. I am confused.

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+Depending on the situation, checking validity of a contract can be expensive. For example, suppose a function
+is designed to perform a binary search on a sorted list of numbers. A pre-condition for the operation is that
+the list it is given to search is sorted. If the list is large, checking that it is properly sorted is expensive.
+In other words, contracts can impact performance. For this reason, it is desirable for callers to have a way to
```
it might be more explicit to say that the impact is not good. 

impact -> "negatively impact" or "impede" or "degrade"

katyhuff commented on 02a-dbc.md:
```diff
@@ -0,0 +1,153 @@
+either by setting an environment variable, DISABLE_CONTRACTS or by a call to contracts.disable_all() **before**
+any @contracts are processed by the Python interpreter. This allows developers to keep the checks in place while
+they are developing code and then disable them once they are sure the code is working as expected.
+
```
This would be a good spot to explain when is a good time to write these contracts. Before you write the function code? while you write the function code? after you write the function code?

markcmiller86 commented:
I took care of everything except the re-numbering, re-positioning. Wondering if you wouldn't mind handling that after a merge? I'm squeamish about too many changes to existing filenames.

markcmiller86 commented:
BTW...no worries about timing. Its fine if you need to ponder more where/how this fits.

markcmiller86 commented:
ping

katyhuff commented:
Hi Mark, this branch has conflicts that need to be resolved before it is elligible for merge. Conflicting files include index.html and index.md. There are instructions for handling conflicts on the command line linked above. Please let me know if you have questions.

markcmiller86 commented:
Maybe I am being dense but I cannot find "...the command line linked above...".

katyhuff commented:
I guess it appears below, once this comment is posted (it's above where I typed my comment, sorry).  


I'll try again. There are instructions for resolving the conflict on the command line, those instructions are linked (actually below).

Here is a screen shot: 
![image](https://user-images.githubusercontent.com/393899/29379465-2f522e5c-8288-11e7-8814-10987e2cb13d.png)

markcmiller86 commented:
Ok, thanks. Will try to resolve later today. Sorry I didn't catch this in orig. PR submission

markcmiller86 commented:
I think I've corrected things. That said, when did an upstream pull, I got a bunch of differences that I think had to do with a change in the directory organization for the newer lesson template. I think I adjusted things correctly for the newer template and it appears there are now no conflicts.

I think its good to go now. Sorry I let this languish for as long as I did. Wasn't aware it was in need of more attention from me.

markcmiller86 commented:
ping

markcmiller86 commented:
ping

markcmiller86 commented:
Happened to be hear looking at something. Just pinging again.

katyhuff commented:
Thanks! Sorry I missed this for so long.

markcmiller86 commented:
No worries. Glad I stumbled back into it ;)
